File: cpp/904. Fruit Into Baskets.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Sliding Window, Time: O(n), Space: O(1)
public:
    int totalFruit(vector<int>& tree) {
        int n = tree.size();       
        int l = 0;
        unordered_map<int, int> counter;
        int ans = 0;
        for(int r = 0; r < n; r++) {
            counter[tree[r]] += 1;
            while(counter.size() > 2) {
                counter[tree[l]] -= 1;
                if(counter[tree[l]] == 0) counter.erase(tree[l]);
                l += 1;
            }
            ans = max(ans, r-l+1);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi tree; int ans;

    tree = {1,2,1};
    ans = solution.totalFruit(tree);
    assert(ans==3);
    
    tree = {0,1,2,2};
    ans = solution.totalFruit(tree);
    assert(ans==3);

    tree = {1,2,3,2,2};
    ans = solution.totalFruit(tree);
    assert(ans==4);

    tree = {3,3,3,1,2,1,1,2,3,3,4};
    ans = solution.totalFruit(tree);
    assert(ans==5);
}

File: python/904-Fruit Into Baskets.py
from typing import List
from collections import Counter

class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        "Sliding Window, Time: O(n), Space: O(1)"
        n = len(tree)
        counter = Counter()
        l, result = 0, 0
        for r in range(n):
            counter[tree[r]] += 1
            while len(counter) > 2:
                counter[tree[l]] -= 1
                if counter[tree[l]] <=0: del counter[tree[l]]
                l += 1
            result = max(result, r-l+1)
        return result

tree = [1,2,1]
result = Solution().totalFruit(tree)
assert result == 3

tree = [0,1,2,2]
result = Solution().totalFruit(tree)
assert result == 3

tree = [1,2,3,2,2]
result = Solution().totalFruit(tree)
assert result == 4

tree = [3,3,3,1,2,1,1,2,3,3,4]
result = Solution().totalFruit(tree)
assert result == 5